### PR TITLE
[new scheduler] Allocate resources for spilled back task to a local view of the remote node

### LINKED ISSUE #9482: [New scheduler] Pass test_advanced_3::test_load_balancing_with_dependencies
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

If we cannot run your script, we cannot fix your issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

There is a race condition in the new scheduler involving "light" heartbeats, i.e. nodes don't report their local resource availability if it is the same as in the last heartbeat. Example with 1 CPU:

1. Node A receives node B's availability: 1/1 CPU.
2. Node A receives resource request R, spills R back to node B, updates B's availability to 0/1 CPU.
3. Node B does not queue R. This can happen either because the owner of T no longer needs the worker lease (in which case B never sees the request) or because B spills back R to a different node.
4. Node B's availability is still 1/1 CPU, so it never sends another heartbeat. A believes B's availability is still 0/1 CPU.

This PR is the simplest solution I could think of that keeps steps 1-3 the same. I think this is desirable because step 2 lets A be precise about how many tasks to spill back to B. There are likely other tests that check for this behavior.

This PR modifies the resource accounting done at each node so that there is a "last reported" copy and a "local view" copy. When node A spills back to B, A only reflects this in its local view of B's resources. On the next heartbeat tick, A resets its local view of every other node to the corresponding "last reported" view.

This modification guarantees that A will eventually reset B's availability to 1/1 after step 4.

## Related issue number

Closes #9482.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ericl** [comment]
Another approach is that the local modifications could be made to a "temporary copy" of the node resources, which is discarded after each heartbeat update. I think this solution is simpler in that it avoids any changes to the distributed protocol, what do you think?

**Comment by stephanie-wang** [comment]
> Another approach is that the local modifications could be made to a "temporary copy" of the node resources, which is discarded after each heartbeat update. I think this solution is simpler in that it avoids any changes to the distributed protocol, what do you think?

Hmm yeah, that makes sense. I agree it's simpler.

Unfortunately Travis still seemed to fail on this PR build, so I'm wondering if there's something else going on. Trying a version with light heartbeats disabled to see if there's any difference.

**Comment by stephanie-wang** [comment]
Note that this PR also contains some cleanups, mostly removing methods that were never called. I included this to reduce the chance that a future change modifies the wrong resource view, now that we have a "last reported" vs a "local" view.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.h [hunk 1]
```diff
@@ -413,6 +352,61 @@ class ClusterResourceScheduler {
...
+    /// The resource information according to the last heartbeat reported by
+    /// this node.
+    /// NOTE(swang): For the local node, this field should be ignored because
+    /// we do not receive heartbeats from ourselves and the local view is
+    /// therefore always the most up-to-date.
+    NodeResources last_reported;
+    /// Our local view of the remote node's resources. This may be dirty
+    /// because it includes any resource requests that we allocated to this
+    /// node through spillback since our last heartbeat tick.
+    NodeResources local_view;
```
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.h:370]: Minor, but we could defensively make these private fields and add accessors that make it harder to mess up, e.g.,

ResetResources()  // updates both reported and local view
UpdateLocal()  // updates local only
GetLocal() const  // get local read view

File: src/ray/raylet/scheduling/cluster_resource_scheduler.h [hunk 2]
```diff
@@ -413,6 +352,61 @@ class ClusterResourceScheduler {
...
+
+    /// The resource information according to the last heartbeat reported by
+    /// this node.
+    /// NOTE(swang): For the local node, this field should be ignored because
+    /// we do not receive heartbeats from ourselves and the local view is
+    /// therefore always the most up-to-date.
+    NodeResources last_reported;
+    /// Our local view of the remote node's resources. This may be dirty
+    /// because it includes any resource requests that we allocated to this
+    /// node through spillback since our last heartbeat tick.
```
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.h:369]: Add that we overwrite local_view to last_reported on each heartbeat (to handle light heartbeats)

**Review by ericl — APPROVED** [review]
LGTM, a suggestion on improving maintainability

**Pull request merged**